src/client/fuse_ll.cc

```diff
@@ -612,9 +612,12 @@ static void fuse_ll_read(fuse_req_t req, fuse_ino_t ino, size_t size, off_t off,
   Fh *fh = reinterpret_cast<Fh*>(fi->fh);
   bufferlist bl;
   int r = cfuse->client->ll_read(fh, off, size, &bl);
-  if (r >= 0)
-    fuse_reply_buf(req, bl.c_str(), bl.length());
-  else
+  if (r >= 0) {
+    vector<iovec> iov;
+    bl.prepare_iov(&iov);
+    iov.insert(iov.begin(), {0}); // the first one is reserved for fuse_out_header
+    fuse_reply_iov(req, &iov[0], iov.size());
+  } else
```

Comment:
this change is good. It avoid calling buffer::list::c_str(), which may rebuild buffer.

Reply:
Are you sure? prepare_iov does this:
```
      for (auto& p : _buffers) {
        (*piov)[n].iov_base = (void *)p.c_str();
        (*piov)[n].iov_len = p.length();
        ++n;
      }
```
Does that c_str() call copy the buffer to ensure that it's NULL terminated, or does the buffers_t type store the data already NULL terminated?

Reply:
neither.  buffer::ptr:c_str() does not do copy. it does not return null terminated str

Reply:
Yeah, Yan is right. The memories in _buffers will be freed when the bl is being destructed.

Reply:
Ugh, so we have a c_str function that doesn't actually return a NULL terminated buffer? That seems confusing. Oh well...ok.

Reply:
This should be okay, since we have the str len at the same in the struct iovec.

Reply:
Yeah, not really a problem with this patch. It's more just me wondering if there are callers of that c_str() method that actually use it like a C string. One would hope not, but the fact that it doesn't actually return a null-terminated string seems like a great way to subtly introduce buffer overruns.

Reply:
Yeah, that's right.